Implement `SelectedValue` and `SelectedValueBinding` for SelectingItemsControls

amwx opened the pull request:
## What does the pull request do?
Adds `SelectedValueBinding` and `SelectedValue` properties to `SelectingItemsControl`. It was implemented similar to how AutoCompleteBox works using an `IBinding` to evaluate the value given a specific item as a data context. Behavior should be otherwise similar to UWP/WPF.

```Xaml
<ListBox Items="{Binding Items}"
             DisplayMemberBinding="{Binding Name}"
             SelectedValue="{Binding SelectedValue}"
             SelectedValueBinding="{Binding Other}"
             SelectedItem="{Binding SelectedItem}"
             Width="300" Height="300"
             Name="LB" />
```

```C#
public class TestClass
{
    public string Name { get; set; }

    public string Other { get; set; }
}
```

![image](https://user-images.githubusercontent.com/40413319/216470199-6a1390af-7e5a-40eb-a3bd-26ceba7b8e81.png)


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

Closes #4718

amwx committed:
Implement SelectedValue

amwx committed:
Add Tests

amwx committed:
Missing xml docs

amwx committed:
Merge branch 'master' into feat/SelectedValue_SelectedValueBinding

amwx commented on src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:
```diff
@@ -1122,6 +1333,54 @@ public Optional<int> SelectedIndex
+        /// <summary>
+        /// Helper class for evaluating a binding from an Item and IBinding instance
+        /// </summary>
+        private class BindingHelper : StyledElement
```
`BindingHelper` class here was adapted from what AutoCompleteBox does, but it requires at least a StyledElement for the DataContext property and at the very least an AvaloniaObject for the property system. Which this has the downside of rigging up a bunch of "backend" stuff when setting the data context and evaluating the binding, which may become a perf issue when scanning a large items source.

Out of curiosity, would it be (at least theoretically) possible to add an extension to evaluate a binding without needing to go through the property system? Something like:

```diff
public class BindingOperations
{
+    public object EvaluateOneTime(IBinding binding, object dataContext)
}
```
Allowing you to specify the binding and the data context and could produce a one time value without the entire property system being invoked. 
There's also other use cases for this type of thing, especially since we're trying to avoid string paths like WPF/UWP which require reflection, and we could get rid of these helper classes in AutoCompleteBox and here. 

Just a question/thought. Not sure if this is feasible though, I don't know enough about the binding internals.

maxkatz6 replied:
I believe EvaluateOneTime can be an internal method at the very least. And used in both controls.
About internal implementation details @grokys might have an idea, as it was planned to refactor bindings system in the future. It also can be made public at some point.

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029683-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

maxkatz6 commented on src/Avalonia.Controls/Primitives/SelectingItemsControl.cs:
```diff
@@ -209,6 +225,27 @@ public int SelectedIndex
+        /// <see cref="SelectedValue"/> property
+        /// </summary>
+        [AssignBinding]
+        public IBinding? SelectedValueBinding
```
It would be great if this property supported compiled bindings without need to specify x:DataType.

I hope [InheritDataTypeFromItems](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ItemsControl.cs#L171) will just work here. And double checked if it works.

amwx replied:
Oh, I completely missed that being added. Will add that attribute and test it works

amwx committed:
Merge branch 'master' into feat/SelectedValue_SelectedValueBinding

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029757-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

amwx committed:
Add `InheritDataTypeFromItems` to SelectedValueBinding

amwx committed:
Code cleanup

amwx committed:
InheritDataTypeFromItems for DisplayMemberBinding & cleanup ItemsControl

amwx commented:
@maxkatz6 Ok, added `InheritDataTypeFromItems` to `SelectedValueBinding` and I also added it to `DisplayMemberBinding` as it wasn't used there either. Can confirm it works:

![image](https://user-images.githubusercontent.com/40413319/217670030-831facbd-618d-459b-90cf-7fe7c15d61c4.png)

avaloniaui-team commented:
You can test this PR using the following package version. `11.0.999-cibuild0029762-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]
